### PR TITLE
Add `DGraphFin` dynamic graph dataset

### PR DESCRIPTION
**Pull request by EdisonLeeeee**
This PR is a re-opened one of https://github.com/pyg-team/pytorch_geometric/pull/5466. The major difference it that this PR enforces manual download and just provides a process script for the dataset.

+ Home page:https://dgraph.xinye.com/introduction
+ Paper: https://openreview.net/forum?id=2rQPxsmjKF&referrer=%5Bthe%20
+ Baselines: https://github.com/DGraphXinye/DGraphFin_baseline

## Usage
```python
from torch_geometric.datasets import DGraphFin
dataset = DGraphFin("data/DGraphFin")
data = dataset[0]
>>> data
Data(x=[3700550, 17], edge_index=[2, 4300999], edge_attr=[4300999], y=[3700550], edge_time=[4300999], train_mask=[3700550], val_mask=[3700550], test_mask=[3700550])
```

For now, I used `Data` instead of `TemporalData` for storing this dataset. 

cc: @Padarn @rusty1s

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ## [2.2.0] - 2022-MM-DD
 ### Added
+
```
Comment (wsad1) [inline CHANGELOG.md:8]: I guess you meant to add a changelog.

File: torch_geometric/datasets/dgraph.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    @property
+    def processed_file_names(self) -> str:
+        return 'data.pt'
+
+    @property
+    def num_classes(self) -> int:
+        return 2
+
+    def process(self):
+        extract_zip(self.raw_paths[0], self.raw_dir, log=False)
```
Comment (wsad1) [inline torch_geometric/datasets/dgraph.py:74]: move this to `download` like done in other datasets.

File: torch_geometric/datasets/dgraph.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    """
+
+    url = "https://dgraph.xinye.com"
+
+    def __init__(self, root: str, transform: Optional[Callable] = None,
+                 pre_transform: Optional[Callable] = None):
+        super().__init__(root, transform, pre_transform)
+        self.data, self.slices = torch.load(self.processed_paths[0])
+
+    def download(self):
```
Comment (wsad1) [inline torch_geometric/datasets/dgraph.py:56]: Any reason we don't call `download_url` here?

File: torch_geometric/datasets/dgraph.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+class DGraphFin(InMemoryDataset):
+    r"""The DGraphFin networks from the
+    `"DGraph: A Large-Scale Financial Dataset for Graph Anomaly Detection"
+    <https://arxiv.org/abs/2207.03579>`_ paper.
+    It is a directed, unweighted dynamic graph consisting of millions of
+    nodes and edges, representing a realistic user-to-user social network
+    in financial industry.
+    Node represents a Finvolution user, and an edge from one
+    user to another means that the user regards the other user
+    as the emergency contact person.
```
Comment (wsad1) [inline torch_geometric/datasets/dgraph.py:20]: mention here that the edges have a time attribute, and the granularity of that attribute.

**Review by wsad1 — COMMENTED** [review]
Mostly looks good. Left some initial comments.

File: CHANGELOG.md (hunk 1 above)
Comment (EdisonLeeeee) [reply CHANGELOG.md:8]: It had been added. There was a conflict with the master branch and I removed it temporarily.

File: torch_geometric/datasets/dgraph.py (hunk 3 above)
Comment (EdisonLeeeee) [reply torch_geometric/datasets/dgraph.py:56]: See https://github.com/pyg-team/pytorch_geometric/pull/5466#issuecomment-1251990418. This dataset can only be downloaded manually by users.

**Review by wsad1 — APPROVED** [review]
Thanks for adding this.

**Comment by EdisonLeeeee** [comment]
Shall add an example for the use of this dataset?

**Comment by EdisonLeeeee** [comment]
Let's leave the example as the follow-up PR if it is necessary. I think this PR is ready to merge.

**Review by Padarn — APPROVED** [review]

File: CHANGELOG.md [hunk 5]
```diff
@@ -5,6 +5,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
...
```
Comment (rusty1s) [inline CHANGELOG.md:8]:
```suggestion
```

**Review by rusty1s — APPROVED** [review]
Thank you!

File: torch_geometric/datasets/dgraph.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+    def process(self):
+        extract_zip(self.raw_paths[0], self.raw_dir, log=False)
+        path = os.path.join(self.raw_dir, "dgraphfin.npz")
+
+        with np.load(path) as loader:
+            x = torch.from_numpy(loader['x']).to(torch.float)
+            y = torch.from_numpy(loader['y']).to(torch.long)
+            edge_index = torch.from_numpy(loader['edge_index']).to(torch.long)
+            edge_attr = torch.from_numpy(loader['edge_type']).to(torch.long)
```
Comment (rusty1s) [inline torch_geometric/datasets/dgraph.py:83]:
```suggestion
            edge_type = torch.from_numpy(loader['edge_type']).to(torch.long)
```

File: torch_geometric/datasets/dgraph.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+            edge_time = torch.from_numpy(loader['edge_timestamp']).to(
+                torch.long)
+            train_nodes = torch.from_numpy(loader['train_mask']).to(torch.long)
+            val_nodes = torch.from_numpy(loader['valid_mask']).to(torch.long)
+            test_nodes = torch.from_numpy(loader['test_mask']).to(torch.long)
+
+            train_mask = index_to_mask(train_nodes, size=x.size(0))
+            val_mask = index_to_mask(val_nodes, size=x.size(0))
+            test_mask = index_to_mask(test_nodes, size=x.size(0))
+            data = Data(x=x, edge_index=edge_index.t(), edge_attr=edge_attr,
```
Comment (rusty1s) [inline torch_geometric/datasets/dgraph.py:93]:
```suggestion
            data = Data(x=x, edge_index=edge_index.t(), edge_type=edge_type,
```

**Pull request merged**